judo/utils/normalization.py

```diff
@@ -0,0 +1,208 @@
+        return x
+
+
+class MinMaxNormalizer(Normalizer):
+    """Normalizer that uses min and max values to scale the data to the range [-1, 1]."""
+
+    def __init__(self, dim: int, min: np.ndarray, max: np.ndarray, eps: float = 1e-8) -> None:
+        """Initialize the normalizer.
+
+        Args:
+            dim: Dimension of the data.
+            min: The minimum values of the data.
+            max: The maximum values of the data.
+            eps: Small value to prevent division by zero.
+        """
+        super().__init__(dim)
+        self.min = min
+        self.max = max
+        self.eps = eps
+
+        # Only normalize the dimensions that are not -inf or inf
+        self.norm_dims = np.where((self.min != -np.inf) & (self.max != np.inf))[0]
+
+    def normalize(self, x: np.ndarray) -> np.ndarray:
+        """Normalize the data."""
+        x_normalized = x.copy()
+        x_normalized[..., self.norm_dims] = (
+            2
+            * (x[..., self.norm_dims] - self.min[self.norm_dims])
+            / (self.max[self.norm_dims] - self.min[self.norm_dims])
+            - 1
+        )
+        return x_normalized
+
+    def denormalize(self, x: np.ndarray) -> np.ndarray:
+        """Denormalize the data."""
+        x_denormalized = x.copy()
+        x_denormalized[..., self.norm_dims] = (x[..., self.norm_dims] + 1) * (
+            self.max[self.norm_dims] - self.min[self.norm_dims]
+        ) / 2 + self.min[self.norm_dims]
+        return x_denormalized
+
+
+class RunningMeanStdNormalizer(Normalizer):
+    """Normalizer that uses running statistics (mean and std) to scale the data."""
+
+    def __init__(
+        self,
+        dim: int,
+        init_std: float = 1.0,
+        min_std: float = 1e-6,
+        max_std: float = 1e6,
+        eps: float = 1e-8,
+    ) -> None:
+        """Initialize the normalizer.
+
+        Args:
+            dim: Dimension of the data.
+            init_std: The initial standard deviation for the running statistics.
+            min_std: The minimum standard deviation.
+            max_std: The maximum standard deviation.
+            eps: Small value to prevent division by zero.
+        """
+        super().__init__(dim)
+        self.eps = eps
+        self.min_std = min_std
+        self.max_std = max_std
+
+        # Running statistics
+        self.count = 0
+        self.mean = np.zeros(dim)
+        self.std = np.ones(dim) * init_std
+
+        # For Welford's online algorithm. M2: sum of squares of differences from the current mean
+        self.M2 = np.zeros(dim)
+
+    def update(self, x: np.ndarray) -> None:
+        """Update the running statistics.
+
+        Args:
+            x: The data to update the running statistics with. Shape=(batch_size, dim).
+        """
+        assert x.ndim == 2, f"Expected 2D array (batch_size, dim), but got {x.ndim}D array {x.shape}"
+        assert x.shape[1] == self.dim, f"Expected dimension {self.dim}, but got {x.shape[1]}"
```

Comment:
is it actually necessary for the normalizer inputs to be 2D? i know you call `axis=0` below, but is there a way to generalize the dimension? just makes things potentially cleaner and future-proofed.